pkg/zeek/varint.go

```diff
@@ -0,0 +1,43 @@
+package zeek
+
+// These functions are like varint but already know their size because
+// zval encoded it for them.
+
+func decodeUint(b []byte) uint64 {
+	n := len(b)
+	u64 := uint64(0)
+	for n > 0 {
+		n--
+		u64 <<= 8
+		u64 |= uint64(b[n])
+	}
+	return u64
+}
+
+func encodeUint(dst []byte, u64 uint64) int {
+	n := 0
+	for u64 != 0 {
+		dst[n] = byte(u64)
+		u64 >>= 8
+		n++
+	}
+	return n
+}
+
+func decodeInt(b []byte) int64 {
+	u64 := decodeUint(b)
+	if u64&1 != 0 {
+		return -int64(u64 >> 1)
+	}
+	return int64(u64 >> 1)
+}
+
+func encodeInt(dst []byte, i int64) int {
+	var u64 uint64
+	if i >= 0 {
+		u64 = uint64(i) << 1
+	} else {
+		u64 = uint64(-i)<<1 | 1
+	}
```

Comment:
This encodes math.MinInt64 as "negative zero" in the sense that
```go
	buf := make([]byte, 8)
	n := encodeInt(buf, math.MinInt64)
	y := decodeInt(buf[0:n])
	fmt.Println(y)
```
prints a zero.

The logic in encoding/binary.Putvarint handles it correctly. I'd just copy it.

Reply:
Oh yeah good catch.  Easy to fix right?
```
func decodeInt(b []byte) int64 {
	u64 := decodeUint(b)
	if u64&1 != 0 {
		u64 >>= 1
		if u64 == 0 {
			return math.MinInt64
		}
		return -int64(u64 >> 1)
	}
	return int64(u64 >> 1)
}
```